chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,434 @@
+import chainer
+from chainer.backends import cuda
+from chainer.functions.array.broadcast import broadcast_to
+from chainer.functions.array.cast import cast
+from chainer.functions.array.concat import concat
+from chainer.functions.array.expand_dims import expand_dims
+from chainer.functions.array.reshape import reshape
+from chainer.functions.array.stack import stack
+from chainer.functions.array.transpose import transpose
+from chainer.functions.math.average import average
+from chainer.functions.math.basic_math import absolute
+from chainer.functions.math.maximum import maximum
+from chainer.functions.math.sum import sum as c_sum
+
+import warnings
+
+
+class DiscriminativeMarginBasedClusteringLoss(object):
+
+    """Discriminative margin-based clustering loss function
+
+    This is the implementation of the following paper:
+    https://arxiv.org/abs/1802.05591
+
+    In segmentation, one of the biggest problem is to have noise at the output
+    of a trained network.
+    For cross-entropy based approaches, if the pixel value is wrong,
+    the loss value will be same independent from the wrong pixel's location.
+    Even though the network gives wrong pixel output, it is desirable
+    to have it as close as possible to the original position.
+    By applying discriminative loss function, groups of segmentation instances
+    can be moved together.
+    This loss function calculates the following three parameters:
+        - Variance Loss:
```

Comment:
Insert a blank line before listing and remove the indentation of each item (these indentations cause doc build error). It should look like:

```
    This loss function calculates the following three parameters:

    - Variance loss:
      Loss to penalize...
    - Distance loss:
      Loss to penalize...
    - Regularization loss:
      Small regularization...

    Args:
        ...
```